### PR TITLE
Exclude target directory from Time Machine

### LINKED ISSUE #3884: Cargo's temp files cause backup churn on macOS
macOS automatically makes hourly backups. Cargo's build temp files are large and frequently changing, causing gigabytes of redundant data backed up every day. 

Because the temp files are in multiple project-relative locations, it's not practical for users to manually exclude all the target folders from backup.

Mac applications generally avoid this problem by using system-designed caches directory, but the issue for it (#1615) seems to be in a stalemate.

So I'm proposing an alternative quick fix that doesn't require changing any paths in Cargo. macOS allows excluding individual directories from backups. I suggest marking `target/{build,deps,incremental}` as excluded. It can be done with:

```objc
[URL setResourceValue: @YES] forKey: NSURLIsExcludedFromBackupKey error:nil];
```

### PR DESCRIPTION
**Pull request by pornel**
Fixes #3884

### CODE REVIEW

**Comment by pornel** [comment]
Tests fail on `http_auth_offered`, which looks like a flaky test unrelated to this change.

**Comment by alexcrichton** [comment]
I don't believe the test is flaky but rather related to the upgrade of libgit2-sys here. Mind either updating the test or leaving the dep as it was?

File: Cargo.toml [hunk 1]
```diff
@@ -53,6 +53,7 @@ url = "1.1"
 
 [target.'cfg(unix)'.dependencies]
 openssl = "0.9"
+core-foundation = { version = "0.4.4", features = ["mac_os_10_7_support"] }
```
Comment (alexcrichton) [inline Cargo.toml]: Could this be in a cfg that only gets pulled in on OSX?

File: src/cargo/ops/cargo_rustc/layout.rs [hunk 2]
```diff
@@ -108,11 +108,41 @@ impl Layout {
...
+    /// Marks files or directories as excluded from Time Machine on macOS
+    ///
+    /// This is recommended to prevent derived/temporary files from bloating backups.
+    fn exclude_from_backups(&self, path: &Path) {
+        use std::ptr;
+        use core_foundation::{url, number, string};
+        use core_foundation::base::TCFType;
+
+        // For compatibility with 10.7 a string is used instead of global kCFURLIsExcludedFromBackupKey
+        let is_excluded_key: Result<string::CFString, _> = "NSURLIsExcludedFromBackupKey".parse();
```
Comment (alexcrichton) [inline src/cargo/ops/cargo_rustc/layout.rs]: I think you can perhaps avoid some indentation issues here with:

```rust
let is_excluded_key: string::CFString = match ... {
    Ok(k) => k,
    Err(_) => return,
};
// ...
```

File: Cargo.toml (hunk 1 above)
Comment (pornel) [inline Cargo.toml]: What's the syntax for this? I've tried `[target.'cfg(macos)'.dependencies]`, but it didn't get picked up.

**Comment by pornel** [comment]
I didn't notice libgit got updated.

I've tried to commit only lines of cargo.lock that had core-foundation, but it broke the file (parse error), so I committed the whole thing.

File: Cargo.toml (hunk 1 above)
Comment (alexcrichton) [inline Cargo.toml]: Ah looks like you found it!

**Comment by alexcrichton** [comment]
Looks like this fails to compile on stable/beta?

File: src/cargo/util/read2.rs [hunk 3]
```diff
@@ -12,7 +12,7 @@ mod imp {
 
     pub fn read2(mut out_pipe: ChildStdout,
                  mut err_pipe: ChildStderr,
-                 mut data: &mut FnMut(bool, &mut Vec<u8>, bool)) -> io::Result<()> {
+                 data: &mut FnMut(bool, &mut Vec<u8>, bool)) -> io::Result<()> {
```
Comment (alexcrichton) [inline src/cargo/util/read2.rs]: stray changes?

File: Cargo.toml [hunk 4]
```diff
@@ -35,7 +35,7 @@ home = "0.3"
 ignore = "^0.2.2"
 jobserver = "0.1.6"
 libc = "0.2"
-libgit2-sys = "0.6"
+libgit2-sys = "=0.6.12"
```
Comment (alexcrichton) [inline Cargo.toml]: We shouldn't commit this `=` change

File: src/cargo/util/read2.rs (hunk 3 above)
Comment (pornel) [inline src/cargo/util/read2.rs]: That's a change on master. Github is displaying diff incorrectly. I'll rebase.

**Comment by pornel** [comment]
Passes after fixes and rebase

**Pull request merged**